README
NAME
    Device::L3GD20 - I2C interface to L3GD20 3 axis GyroScope using
    Device::SMBus

VERSION
    version 0.007

ATTRIBUTES
  I2CBusDevicePath
    this is the device file path for your I2CBus that the L3GD20 is
    connected on e.g. /dev/i2c-1 This must be provided during object
    creation.

  Gyroscope
        $self->Gyroscope->enable();
        $self->Gyroscope->getReading();

    This is a object of Device::Gyroscope::L3GD20

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through github at
    <https://github.com/shantanubhadoria/device-l3gd20/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/shantanubhadoria/device-l3gd20>

      git clone git://github.com/shantanubhadoria/device-l3gd20.git

AUTHOR
    Shantanu Bhadoria <shantanu at cpan dott org>

CONTRIBUTORS
    *   Shantanu <shantanu@cpan.org>

    *   Shantanu Bhadoria <shantanu@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Shantanu Bhadoria.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Device-L3GD20

0.007 2013-10-05T00:46:32
 - Merge branch 'master' of github.com:shantanubhadoria/Device-L3GD20
   [Shantanu]
 - new readings [Shantanu Bhadoria]

0.006 2013-10-05T00:46:00
 - v0.006
   
    - new readings [Shantanu Bhadoria]
    - Renamed getReading to getReadingDegreesPerSecond [Shantanu Bhadoria]
   [Shantanu]
 - new readings [Shantanu Bhadoria]
 - Renamed getReading to getReadingDegreesPerSecond [Shantanu Bhadoria]

0.005 2013-09-29T02:46:07
 - v0.005
   
    - Pod Spelling Fixes [Shantanu] [Shantanu]
 - Pod Spelling Fixes [Shantanu]

0.004 2013-09-27T10:18:18
 - v0.004
   
    - Xt test fixes [Shantanu]
    - Fixed error in L3GD20.pm [Shantanu]
    - Refactored Libraries to give all similar devices a common prefix for
      easy identification as plugins [Shantanu]
    - Merge branch 'master' of github.com:shantanubhadoria/Device-L3GD20
      [Shantanu Bhadoria]
    - minor Changes [Shantanu Bhadoria] [Shantanu]
 - Xt test fixes [Shantanu]
 - Fixed error in L3GD20.pm [Shantanu]
 - Refactored Libraries to give all similar devices a common prefix for
   easy identification as plugins [Shantanu]
 - Merge branch 'master' of github.com:shantanubhadoria/Device-L3GD20
   [Shantanu Bhadoria]
 - minor Changes [Shantanu Bhadoria]

0.003 2013-09-12T09:21:01
 - v0.003
   
    - Added README.pod and Changes [Shantanu] [Shantanu]
 - Added README.pod and Changes [Shantanu]

0.002 2013-09-11T09:13:46
 - pod fix [Shantanu]

0.001 2013-09-11T08:25:26
 - First Release [Shantanu]
 - First Release [Shantanu]
 - first commit [Shantanu Bhadoria]
 -  minor commit [Shantanu Bhadoria]
 - First Commit [Shantanu Bhadoria]



examples/example.pl
use FindBin qw($Bin);
use lib "$Bin/../lib";

use Device::Gyroscope::L3GD20;

my $g = Device::Gyroscope::L3GD20->new(
    I2CBusDevicePath => '/dev/i2c-1',
    XZero            => -26.99,
    YZero            => 7.46,
    ZZero            => -9.07,
);

$g->enable();
use Data::Dumper;
my ( $x,  $y,  $z )  = ( 0, 0, 0 );
my ( $dx, $dy, $dz ) = ( 0, 0, 0 );
my $count = 0;
while () {
    $gyro = $g->getReadingDegreesPerSecond();
    $graw = $g->getRawReading();

    #    print 'Gyroscope(Raw): ' . Dumper $g->getRawReading();
    #    print "Gyroscope: $gyro->{x}\t$gyro->{y}\t$gyro->{z}\n" ;
    $dx += $gyro->{x} * $g->timeDrift;
    $dy += $gyro->{y} * $g->timeDrift;
    $dz += $gyro->{z} * $g->timeDrift;
    print "Gyroscope: $gyro->{x}\t$gyro->{y}\t$gyro->{z}\n";

    #    print "Gyroscope(aggregate): $dx\t$dy\t$dz\n" ;
    #    print "TIME DRIFT: " . $g->timeDrift . "\n" ;

=head2

    $x+= $graw->{x};
    $y+= $graw->{y};
    $z+= $graw->{z};
    $count++;
    print "AVERAGE: " . ($x/$count) . ", " . ($y/$count) . ", " . ($z/$count) . "\n";

=cut

}


lib/Device/L3GD20.pm
use strict;
use warnings;

package Device::L3GD20;

# PODNAME: Device::L3GD20
# ABSTRACT: I2C interface to L3GD20 3 axis GyroScope using Device::SMBus
#
# This file is part of Device-L3GD20
#
# This software is copyright (c) 2013 by Shantanu Bhadoria.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
our $VERSION = '0.007'; # VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

use Device::Gyroscope::L3GD20;


has 'I2CBusDevicePath' => ( is => 'ro', );


has Gyroscope => (
    is         => 'ro',
    isa        => 'Device::Gyroscope::L3GD20',
    lazy_build => 1,
);

sub _build_Gyroscope {
    my ($self) = @_;
    my $obj =
      Device::Gyroscope::L3GD20->new(
        I2CBusDevicePath => $self->I2CBusDevicePath, );
    return $obj;
}

1;

__END__

=pod

=head1 NAME

Device::L3GD20 - I2C interface to L3GD20 3 axis GyroScope using Device::SMBus

=head1 VERSION

version 0.007

=head1 ATTRIBUTES

=head2 I2CBusDevicePath

this is the device file path for your I2CBus that the L3GD20 is connected on e.g. /dev/i2c-1
This must be provided during object creation.

=head2 Gyroscope

    $self->Gyroscope->enable();
    $self->Gyroscope->getReading();

This is a object of L<Device::Gyroscope::L3GD20>

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through github at 
L<https://github.com/shantanubhadoria/device-l3gd20/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/shantanubhadoria/device-l3gd20>

  git clone git://github.com/shantanubhadoria/device-l3gd20.git

=head1 AUTHOR

Shantanu Bhadoria <shantanu at cpan dott org>

=head1 CONTRIBUTORS

=over 4

=item *

Shantanu <shantanu@cpan.org>

=item *

Shantanu Bhadoria <shantanu@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Device/L3GD20/Gyroscope.pm
package Device::L3GD20::Gyroscope;

# PODNAME: Device::L3GD20::Gyroscope
# ABSTRACT: I2C interface to Gyroscope on the L3GD20 using Device::SMBus
#
# This file is part of Device-L3GD20
#
# This software is copyright (c) 2013 by Shantanu Bhadoria.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
our $VERSION = '0.003'; # VERSION

use 5.010;
use Moose;
use POSIX;

extends 'Device::SMBus';

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x6b,
);

# Registers for the Gyroscope
use constant {
    CTRL_REG1 => 0x20,
    CTRL_REG4 => 0x23,
};

# X, Y and Z Axis Gyroscope Data value in 2's complement
use constant {
    OUT_X_H => 0x29,
    OUT_X_L => 0x28,

    OUT_Y_H => 0x2b,
    OUT_Y_L => 0x2a,

    OUT_Z_H => 0x2d,
    OUT_Z_L => 0x2c,
};

use integer
  ; # Use arithmetic right shift instead of unsigned binary right shift with >> 4


sub enable {
    my ($self) = @_;
    $self->writeByteData( CTRL_REG1, 0b00001111 );
    $self->writeByteData( CTRL_REG4, 0b00110000 );
}


sub getRawReading {
    my ($self) = @_;

    return (
        x => (
            $self->_typecast_int_to_int16(
                ( $self->readByteData(OUT_X_H) << 8 ) |
                  $self->readByteData(OUT_X_L)
            )
        ),
        y => (
            $self->_typecast_int_to_int16(
                ( $self->readByteData(OUT_Y_H) << 8 ) |
                  $self->readByteData(OUT_Y_L)
            )
        ),
        z => (
            $self->_typecast_int_to_int16(
                ( $self->readByteData(OUT_Z_H) << 8 ) |
                  $self->readByteData(OUT_Z_L)
            )
        ),
    );
}

sub _typecast_int_to_int16 {
    return unpack 's' => pack 'S' => $_[1];
}

sub calibrate {
    my ($self) = @_;

}

1;

__END__

=pod

=head1 NAME

Device::L3GD20::Gyroscope - I2C interface to Gyroscope on the L3GD20 using Device::SMBus

=head1 VERSION

version 0.003

=head1 METHODS

=head2 enable 

    $self->enable()

Initializes the device, Call this before you start using the device. This function sets up the appropriate default registers.
The Device will not work properly unless you call this function

=head2 getRawReading

    $self->getRawReading()

Return raw readings from accelerometer registers

=head1 AUTHOR

Shantanu Bhadoria <shantanu at cpan dott org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla::Plugin::MakeMaker::Awesome 0.15.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "I2C interface to L3GD20 3 axis GyroScope using Device::SMBus",
  "AUTHOR" => "Shantanu Bhadoria <shantanu at cpan dott org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Device-L3GD20",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Device::L3GD20",
  "PREREQ_PM" => {
    "Device::SMBus" => 0,
    "Math::Trig" => 0,
    "Moose" => 0,
    "POSIX" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.007",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



